{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Multiplicity figure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict, Counter\n",
    "import pandas as pd\n",
    "from scipy import stats as sci_stats\n",
    "from matplotlib import pyplot as pl\n",
    "from matplotlib import cm\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "from glob import glob\n",
    "import matplotlib.gridspec as gridspec\n",
    "from matplotlib.colors import ListedColormap, to_rgba\n",
    "from statsmodels.stats.multitest import fdrcorrection as benjamini_hochberg\n",
    "from matplotlib.patches import Rectangle\n",
    "\n",
    "%matplotlib inline\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plates = ['P1', 'P2', 'P3']\n",
    "plate2env = {'P1': r'YPD 30$\\degree$C', 'P2': r'SC 30$\\degree$C', 'P3': r'SC 37$\\degree$C'}\n",
    "strains = ['diploid', 'alpha', 'a']\n",
    "strains_for_print = {'a': '$MATa$', 'diploid': 'Diploid', 'alpha': r'$MAT\\alpha$'}\n",
    "color_by_strain = {'diploid': '#555555', 'alpha': '#FFB000', 'a': '#648FFF'}\n",
    "fa_gens = [70, 550, 1410, 2640, 3630, 5150, 7530, 10150]\n",
    "seq_gens = [70, 1410, 2640, 5150, 7530, 10150]\n",
    "all_wells = sorted([i.split('/')[-1].split('_')[0] for i in glob('../../Output/WGS/combined_option/processed_well_output/*_processed.tsv')])\n",
    "wells = [w for w in all_wells if w!='P1B03'] #P1B03 excluded because it is a haploid population that diploidized\n",
    "gene_info = pd.read_csv('../accessory_files/yeast_gene_annotations.tsv', delimiter='\\t')\n",
    "gene_info = gene_info[gene_info['featureType']=='ORF'].loc[gene_info['briefDescription'].apply(lambda bd: ('Putative protein' not in bd) and ('Dubious open reading frame' not in bd))]\n",
    "gene_to_start_end = {i[0]: i[1:] for i in gene_info.as_matrix(['Gene_ORF', 'start', 'end'])}\n",
    "orf_sizes = list(gene_info['end']-gene_info['start'])\n",
    "\n",
    "essential_orfs_by_Liu = list(gene_info[gene_info['Essential_by_Liu2015']]['ORF'])\n",
    "essential_orfs_by_Gaiever_not_Liu = [i for i in gene_info[gene_info['Essential_by_Giaever2002']]['ORF'] if i not in essential_orfs_by_Liu]\n",
    "\n",
    "o2g = {i[0]:i[1] for i in gene_info.as_matrix(['ORF', 'Gene_ORF']) if pd.notnull(i[1])}\n",
    "o2g.update({i[0]:i[0] for i in gene_info.as_matrix(['ORF', 'Gene_ORF']) if pd.isnull(i[1])})\n",
    "g2o = {o2g[o]:o for o in o2g}\n",
    "wellinfo = pd.read_csv('../accessory_files/VLTE_by_well_info.csv')[['plate.well', 'contam', 'strain']]\n",
    "wellinfo['plate_well'] = wellinfo['plate.well'].apply(lambda p: p[:2]+p[3:]) #reformatting to match for merge\n",
    "well_to_strain = {i[0]:i[1] for i in wellinfo.as_matrix(['plate_well', 'strain'])}\n",
    "wells_w_ade2_stop_lost = ['P2F07', 'P1C09', 'P1E11', 'P3B10', 'P2B09']\n",
    "cb_pal = sns.color_palette('colorblind')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Loading mutation data for next figures"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Some code for calculating mutational opportunities:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nt2codon = {\n",
    "    'TTT': 'F', 'TTC': 'F',\n",
    "    'TTA': 'L', 'TTG': 'L', 'CTT': 'L', 'CTC': 'L', 'CTA': 'L', 'CTG': 'L',\n",
    "    'TCT': 'S', 'TCC': 'S', 'TCA': 'S', 'TCG': 'S', 'AGT': 'S', 'AGC': 'S',\n",
    "    'TAT': 'Y', 'TAC': 'Y',\n",
    "    'TAA': '*', 'TAG': '*', 'TGA': '*',\n",
    "    'TGT': 'C', 'TGC': 'C',\n",
    "    'TGG': 'W',\n",
    "    'CCT': 'P', 'CCC': 'P', 'CCA': 'P', 'CCG': 'P',\n",
    "    'CAT': 'H', 'CAC': 'H',\n",
    "    'CAA': 'Q', 'CAG': 'Q',\n",
    "    'CGT': 'R', 'CGC': 'R', 'CGA': 'R', 'CGG': 'R', 'AGA': 'R', 'AGG': 'R',\n",
    "    'ATT': 'I', 'ATC': 'I', 'ATA': 'I',\n",
    "    'ATG': 'M',\n",
    "    'ACT': 'T', 'ACC': 'T', 'ACA': 'T', 'ACG': 'T',\n",
    "    'AAT': 'N', 'AAC': 'N',\n",
    "    'AAA': 'K', 'AAG': 'K',\n",
    "    'GTT': 'V', 'GTC': 'V', 'GTA': 'V', 'GTG': 'V',\n",
    "    'GCT': 'A', 'GCC': 'A', 'GCA': 'A', 'GCG': 'A',\n",
    "    'GAT': 'D', 'GAC': 'D',\n",
    "    'GAA': 'E', 'GAG': 'E',\n",
    "    'GGT': 'G', 'GGC': 'G', 'GGA': 'G', 'GGG': 'G'\n",
    "}\n",
    "\n",
    "def get_attrib(row, attrib):\n",
    "    if row['type']=='gene':\n",
    "        if attrib+'=' in row['attributes']:\n",
    "            return row['attributes'].split(attrib+'=')[1].split(';')[0]\n",
    "    return ''\n",
    "\n",
    "def read_fasta(fasta_file):\n",
    "    \"\"\"\n",
    "    Reads a fasta file and returns a dictionary with seqid keys and sequence values\n",
    "    \"\"\"\n",
    "    fd = dict()\n",
    "    with open(fasta_file, 'r') as infile:\n",
    "        for line in infile:\n",
    "            if '>' in line:\n",
    "                current_key = line[1:].strip()\n",
    "                fd[current_key] = ''\n",
    "            else:\n",
    "                fd[current_key] += line.strip()\n",
    "    return fd\n",
    "\n",
    "def reverse_transcribe(seq):\n",
    "    \"\"\"reverse transcribes a dna sequence (does not convert any non-atcg/ATCG characters)\"\"\"\n",
    "    watson_crick = {'A': 'T', 'T': 'A', 'G': 'C', 'C': 'G', 'a': 't', 't': 'a', 'g': 'c', 'c': 'g'}\n",
    "    return ''.join([watson_crick.setdefault(c, c) for c in seq[::-1]])\n",
    "\n",
    "class SeqInfoGetter:\n",
    "    \n",
    "    def __init__(self, gff_file, fasta_file):\n",
    "        gff_cols = ['seqid', 'source', 'type', 'start', 'end', 'score', 'strand', 'phase', 'attributes']\n",
    "        self.gff = pd.read_csv(gff_file, delimiter='\\t', skiprows=1, header=None, names=gff_cols)\n",
    "        self.gff['ORF'] = self.gff.apply(lambda row: get_attrib(row, \"ID\"), axis=1)\n",
    "        self.genes = self.gff[self.gff['ORF']!='']\n",
    "        self.genes['Gene'] = self.genes.apply(lambda row: get_attrib(row, \"gene\"), axis=1)\n",
    "        self.chromo_seqs = read_fasta(fasta_file)\n",
    "    \n",
    "    def get_nt_seq(self, element_name, element_type):\n",
    "        td = self.genes[self.genes[element_type]==element_name]\n",
    "        if len(td) != 1:\n",
    "            print(len(td), 'hits, aborting.')\n",
    "            return None\n",
    "        else:\n",
    "            row = td.iloc[0]\n",
    "            cs = self.chromo_seqs[row['seqid']]\n",
    "            if row['strand'] == '+':\n",
    "                return cs[row['start']-1:row['end']]\n",
    "            else:\n",
    "                return reverse_transcribe(cs[row['start']-1:row['end']])\n",
    "            \n",
    "    def get_aa_seq(self, element_name, element_type):\n",
    "        nt_s = self.get_nt_seq(element_name, element_type)\n",
    "        if nt_s:\n",
    "            aas = ''\n",
    "            for i in range(len(nt_s)//3):\n",
    "                aas += nt2codon[nt_s[i*3:(i+1)*3]]\n",
    "            if len(nt_s) % 3 != 0:\n",
    "                aas += '-leftover->' + nt_s[-1*(len(nt_s) % 3):]\n",
    "            return aas\n",
    "            \n",
    "    def get_mutational_opps(self, element_name, element_type, verbose=False, return_nonsyn_over_all=False):\n",
    "        nt_s = self.get_nt_seq(element_name, element_type)\n",
    "        if nt_s:\n",
    "            if len(nt_s) % 3 != 0:\n",
    "                if verbose:\n",
    "                    print('Warning: seq len not a multiple of 3', element_name)\n",
    "                    print(self.genes[self.genes[element_type]==element_name].iloc[0]['Gene'])\n",
    "                    print(self.get_aa_seq(element_name, element_type))\n",
    "           \n",
    "            syn, nonsyn = 0, 0\n",
    "            for i in range(len(nt_s)//3):\n",
    "                codon_seq = nt_s[i*3:(i+1)*3]\n",
    "                codes_for = nt2codon[codon_seq]\n",
    "                for j in range(3):\n",
    "                    for nt in 'ATCG':\n",
    "                        if nt != codon_seq[j]:\n",
    "                            if nt2codon[codon_seq[:j]+nt+codon_seq[j+1:]] == codes_for:\n",
    "                                syn += 1\n",
    "                            else:\n",
    "                                nonsyn += 1\n",
    "            if return_nonsyn_over_all:\n",
    "                return nonsyn/(syn+nonsyn)\n",
    "            else:\n",
    "                return nonsyn / syn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seqI = SeqInfoGetter('../../Output/WGS/reference/w303_vlte.gff', '../../Output/WGS/reference/w303_vlte.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "orf_lens = {o: len(seqI.get_nt_seq(o, \"ORF\")) for o in seqI.genes['ORF']}\n",
    "orf_mutational_opp_ratios = {o: seqI.get_mutational_opps(o, \"ORF\") for o in seqI.genes['ORF']} # Yeilds % of nonsyn/syn random mutations in each ORF\n",
    "orf_mutational_nonsyn_opps = {o: seqI.get_mutational_opps(o, \"ORF\", return_nonsyn_over_all=True)*orf_lens[o] for o in seqI.genes['ORF']} # Yeilds % of nonsyn/syn random mutations in each ORF\n",
    "total_len = np.sum(list(orf_lens.values()))\n",
    "total_nonsyn_ratio = np.sum([orf_mutational_opp_ratios[o]*orf_lens[o]/total_len for o in orf_lens])\n",
    "total_nonsyn_ratio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def is_snp(row):\n",
    "    if row['mutation_type'] != 'Indel':\n",
    "        # * is given if there is a spanning deletion at this site (so no counts for ref or alt (not a SNP)))\n",
    "        if len(row['REF']) == 1 and len(row['ALT'])==1 and row['ALT'] != '*': \n",
    "            return True\n",
    "    return False\n",
    " \n",
    "def hit_orfs(orf_list, search_list):\n",
    "    for o in str(orf_list).split(';'):\n",
    "        if o in search_list:\n",
    "            return True\n",
    "    return False\n",
    "\n",
    "# by well dataframes with mutations\n",
    "well_dats = dict()\n",
    "for well in wells:\n",
    "    well_dats[well] = pd.read_csv('../../Output/WGS/combined_option/processed_well_output/' + well + '_processed.tsv', delimiter='\\t')\n",
    "    # Exclude from analysis mutations in the 2-micron plasmid and telomeres, and SVs\n",
    "    well_dats[well] = well_dats[well][pd.isnull(well_dats[well]['SVTYPE']) & (well_dats[well]['CHROM']!='2-micron') & (~well_dats[well]['in_telomere'])]\n",
    "    well_dats[well]['is_snp'] = well_dats[well].apply(lambda r: is_snp(r), axis=1)\n",
    "\n",
    "# a dataframe with hits and multiplicity for each ORF in the yeast genome\n",
    "orf_hit_df = pd.read_csv('../../Output/WGS/combined_option/gene_hit_data.tsv', delimiter='\\t')\n",
    "orf_hit_df = orf_hit_df.merge(gene_info[['ORF', 'briefDescription', 'Essential_by_Liu2015', 'Essential_by_Giaever2002', 'start', 'end']], on='ORF', how='left')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Multiplicity fig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# getting how many times each amino acid position is hit\n",
    "# for now just taking the first annotation from the ANN column: \n",
    "# when split by |, the 14th column is the codon position like 54/109\n",
    "aa_hits = defaultdict(set)\n",
    "for well in wells:\n",
    "    td = well_dats[well]\n",
    "    mgs_seen = set()\n",
    "    for entry in td[td['fixed_by_10150'] & pd.notnull(td['ORF_hit'])].as_matrix(['ANN', 'ORF_hit', 'mutation_type', 'CHROM', 'POS', 'REF', 'ALT', 'mutation_group']):\n",
    "        if entry[7] not in mgs_seen:\n",
    "            mgs_seen.add(entry[7])\n",
    "            aa_pos_split = str(entry[0]).split('|')\n",
    "            if len(aa_pos_split) > 13:\n",
    "                if aa_pos_split[13] != '':\n",
    "                    aa_hits[entry[1]+'_'+aa_pos_split[13]].add(well+' '+str(entry[2])+' '+str(entry[3])+' '+str(entry[4])+' ' + '->'.join(entry[5:7]) + ' '+ aa_pos_split[10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Simulating multiplicity by drawing genes to hit for each population,\n",
    "# taking into account the number of hit mutations in each population,\n",
    "# and the lengths of all ORFs in the yeast genome\n",
    "def simulate_gene_hits(well_num_hits, nsamps=1000):\n",
    "    all_m_opps = list(orf_mutational_nonsyn_opps.values())\n",
    "    orf_hits = [0 for o in all_m_opps]\n",
    "    mean_opps = np.mean(all_m_opps)\n",
    "    orf_hit_probs = np.array(all_m_opps)/np.sum(all_m_opps)\n",
    "    multiplicities = []\n",
    "    pops_hit = []\n",
    "    for n in range(nsamps):\n",
    "        hit_table = []\n",
    "        for num_hits in well_num_hits:\n",
    "            hit_table.append(np.random.multinomial(num_hits, orf_hit_probs))\n",
    "        hit_table = np.array(hit_table)\n",
    "        orf_hits = np.sum(hit_table, axis=0)\n",
    "        multiplicities += [mult for mult in list(mean_opps * (orf_hits / np.array(all_m_opps))) if mult != 0] # we do not include orfs with zero hits\n",
    "        pops_hit += list(np.sum(np.clip(hit_table, 0, 1), axis=0))\n",
    "    return multiplicities, pops_hit\n",
    "\n",
    "pop_hits = np.sum(orf_hit_df[wells], axis=0)\n",
    "sim_mult, sim_hits = simulate_gene_hits(pop_hits)\n",
    "\n",
    "pop_hits = np.sum(orf_hit_df[wells], axis=0)\n",
    "sim_mult, sim_hits = simulate_gene_hits(pop_hits)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def simulate_aa_pos_hits(nsamps=100):\n",
    "    ## What I want to do is actually look at the ORFs that are hit and randomize which codon they hit\n",
    "    pops_hit = []\n",
    "    for n in range(nsamps):\n",
    "        aa_sim_hits_dict = defaultdict(set)\n",
    "        for well in wells:\n",
    "            for entry in np.array(orf_hit_df[['ORF', 'size', well]]):\n",
    "                for i in range(entry[2]):\n",
    "                    aa_sim_hits_dict[entry[0]+'_'+str(np.random.randint(entry[1]))].add(well)\n",
    "        pops_hit += [len(i) for i in aa_sim_hits_dict.values()]\n",
    "    return pops_hit\n",
    "\n",
    "sim_aa_hits = simulate_aa_pos_hits()\n",
    "sim_aa_hits += [0]*(np.sum(list(orf_lens.values()))//3-len(sim_aa_hits))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "f, subs = pl.subplots(1, 3, figsize=(7.5, 1.5), dpi=300)\n",
    "pl.subplots_adjust(wspace=0.7)\n",
    "\n",
    "actual_mult = list(orf_hit_df['multiplicity'])\n",
    "actual_mult += [0]*(len(orf_lens)-len(orf_hit_df))\n",
    "max_m = int(np.ceil(max(actual_mult)))\n",
    "subs[0].hist(sim_mult, histtype='step', log=True, bins=[i for i in range(max_m)], cumulative=-1, edgecolor='k', alpha=0.5, label='Null', \n",
    "         weights=np.ones_like(sim_mult)/float(len(sim_mult)))\n",
    "\n",
    "subs[0].hist(actual_mult, histtype='step', log=True, bins=[i for i in range(max_m)], cumulative=-1, lw=1, label='Actual', \n",
    "         weights=np.ones_like(actual_mult)/float(len(actual_mult)))\n",
    "subs[0].set_xlabel('Multiplicity ($m$)', fontsize=9)\n",
    "subs[0].set_ylabel('Fraction of\\nGenes ' + r'$\\geq m$', fontsize=9)\n",
    "subs[0].set_ylim([0.5/len(orf_lens), 1.1])\n",
    "\n",
    "actual_hits = list(orf_hit_df['pops_hit'])\n",
    "actual_hits += [0]*(len(orf_lens)-len(orf_hit_df))\n",
    "max_m = int(np.ceil(max(actual_hits)))\n",
    "subs[1].hist(sim_hits, histtype='step', log=True, bins=[i for i in range(max_m)], cumulative=-1, edgecolor='k', alpha=0.5, label='Null', \n",
    "         weights=np.ones_like(sim_hits)/float(len(sim_hits)))\n",
    "\n",
    "for i in range(5,8):\n",
    "    print('Prob of getting', i, 'pop hits or more:', len([j for j in sim_hits if j>=i])/len(sim_hits))\n",
    "\n",
    "subs[1].hist(actual_hits, histtype='step', log=True, bins=[i for i in range(max_m)], cumulative=-1, lw=1, label='Actual', \n",
    "         weights=np.ones_like(actual_hits)/float(len(actual_hits)))\n",
    "subs[1].set_xlabel('Populations hit ($PH$)', fontsize=9)\n",
    "subs[1].set_ylabel('Fraction of\\nGenes ' + r'$\\geq PH$', fontsize=9)\n",
    "subs[1].set_ylim([0.5/len(orf_lens), 1.1])\n",
    "\n",
    "actual_aa_hits = [len(aa_hits[a]) for a in aa_hits]\n",
    "actual_aa_hits += [0]*(np.sum(list(orf_lens.values()))//3-len(actual_aa_hits))\n",
    "max_m = int(np.ceil(max(actual_aa_hits)))\n",
    "subs[2].hist(sim_aa_hits, histtype='step', log=True, bins=[i for i in range(max_m)], cumulative=-1, edgecolor='k', alpha=0.5, label='Null', \n",
    "         weights=np.ones_like(sim_aa_hits)/float(len(sim_aa_hits)))\n",
    "\n",
    "subs[2].hist(actual_aa_hits, histtype='step', log=True, bins=[i for i in range(max_m)], cumulative=-1, lw=1, label='Actual', \n",
    "         weights=np.ones_like(actual_aa_hits)/float(len(actual_aa_hits)))\n",
    "subs[2].set_xlabel('Populations hit ($PH$)', fontsize=9)\n",
    "subs[2].set_ylabel('Fraction of\\nAA sites ' + r'$\\geq PH$', fontsize=9)\n",
    "\n",
    "\n",
    "subs[2].legend(frameon=False, fontsize=7)\n",
    "lets = 'ABC'\n",
    "for i in range(3):\n",
    "    subs[i].annotate(lets[i], fontsize=12, xy=(-0.65, 1.1), xycoords=\"axes fraction\", horizontalalignment=\"center\")\n",
    "        \n",
    "sns.despine()\n",
    "\n",
    "f.savefig('../../Output/Figs/Figure6_multiplicity.png', background='transparent', bbox_inches='tight', pad_inches=0.1)\n",
    "f.savefig('../../Output/Figs/Figure6_multiplicity.svg', background='transparent', bbox_inches='tight', pad_inches=0.1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Note that a lot of these are indels that may be hypermutable due to repetitive regions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "orf_hit_nums = {i[0]:i[1] for i in np.array(orf_hit_df[['ORF', 'num_hits']])}\n",
    "orf_codon_nums = {i[0]:i[1]//3 for i in np.array(orf_hit_df[['ORF', 'size']])}\n",
    "\n",
    "for aa in aa_hits:\n",
    "    if len(aa_hits[aa])>2:\n",
    "        print(aa, o2g.get(aa.split('_')[0], 'NA'), len(aa_hits[aa]), len(set([a.split(' ')[0] for a in aa_hits[aa]])))\n",
    "        print(orf_hit_nums[aa.split('_')[0]], orf_codon_nums[aa.split('_')[0]])\n",
    "        print('P value:', (1-sci_stats.binom.cdf(len(aa_hits[aa])-1, orf_hit_nums[aa.split('_')[0]], (1/orf_codon_nums[aa.split('_')[0]])))*orf_codon_nums[aa.split('_')[0]])\n",
    "        for h in aa_hits[aa]:\n",
    "            print(h)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "milo_py37",
   "language": "python",
   "name": "milo_py37"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
