{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict, Counter\n",
    "import pandas as pd\n",
    "from scipy import stats as sci_stats\n",
    "from matplotlib import pyplot as pl\n",
    "from matplotlib import cm\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "from glob import glob\n",
    "\n",
    "%matplotlib inline\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "plates = ['P1', 'P2', 'P3']\n",
    "plate2env = {'P1': 'YPD 30C', 'P2': 'SC 30C', 'P3': 'SC 37C'}\n",
    "strains = ['diploid', 'alpha', 'a']\n",
    "strains_for_print = {'a': '$MATa$', 'diploid': 'Diploid', 'alpha': r'$MAT\\alpha$'}\n",
    "color_by_strain = {'diploid': '#555555', 'alpha': '#FFB000', 'a': '#648FFF'}\n",
    "fa_gens = [70, 550, 1410, 2640, 3630, 5150, 7530, 10150]\n",
    "seq_gens = [70, 1410, 2640, 5150, 7530, 10150]\n",
    "# This dictionary changes the recorded generation numbers to the correct generation numbers\n",
    "# Since P3 only does 8 gens/day it is very different, the other differences are due to little recording errors and unfreezing (~5 generations)\n",
    "gen_fixer = {70.0: {'P1': 70.0, 'P2': 70.0, 'P3': 56.0},\n",
    "             550.0: {'P1': 545.0, 'P2': 545.0, 'P3': 437.0},\n",
    "             1410.0: {'P1': 1395.0, 'P2': 1395.0, 'P3': 1119.0},\n",
    "             2640.0: {'P1': 2615.0, 'P2': 2615.0, 'P3': 2089.0},\n",
    "             3630.0: {'P1': 3620.0, 'P2': 3620.0, 'P3': 2894.0},\n",
    "             5150.0: {'P1': 5115.0, 'P2': 5115.0, 'P3': 4095.0},\n",
    "             7530.0: {'P1': 7460.0, 'P2': 7470.0, 'P3': 5984.0},\n",
    "             10150.0: {'P1': 10070.0, 'P2': 10080.0, 'P3': 8019.0}}\n",
    "wells = sorted([i.split('/')[-1].split('_')[0] for i in glob('../Output/WGS/combined_option/processed_well_output/*_processed.tsv')])\n",
    "gene_info = pd.read_csv('accessory_files/yeast_gene_annotations.tsv', delimiter='\\t')\n",
    "gene_info = gene_info[gene_info['featureType']=='ORF'].loc[gene_info['briefDescription'].apply(lambda bd: ('Putative protein' not in bd) and ('Dubious open reading frame' not in bd))]\n",
    "gene_to_start_end = {i[0]: i[1:] for i in gene_info.as_matrix(['Gene_ORF', 'start', 'end'])}\n",
    "orf_sizes = list(gene_info['end']-gene_info['start'])\n",
    "essential_orfs = list(gene_info[gene_info['Essential']]['ORF'])\n",
    "o2g = {i[0]:i[1] for i in gene_info.as_matrix(['ORF', 'Gene_ORF']) if pd.notnull(i[1])}\n",
    "o2g.update({i[0]:i[0] for i in gene_info.as_matrix(['ORF', 'Gene_ORF']) if pd.isnull(i[1])})\n",
    "g2o = {o2g[o]:o for o in o2g}\n",
    "wellinfo = pd.read_csv('accessory_files/VLTE_by_well_info.csv')[['plate.well', 'contam', 'strain']]\n",
    "wellinfo['plate_well'] = wellinfo['plate.well'].apply(lambda p: p[:2]+p[3:]) #reformatting to match for merge\n",
    "well_to_strain = {i[0]:i[1] for i in wellinfo.as_matrix(['plate_well', 'strain'])}\n",
    "wells_w_ade2_stop_lost = ['P2F07', 'P1C09', 'P1E11', 'P3B10', 'P2B09']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "ids_to_strains_fixed = {s: Counter() for s in strains}\n",
    "for well in wells:\n",
    "    td = pd.read_csv('../Output/WGS/combined_option/well_output/'+well + '_full.tsv', delimiter='\\t')\n",
    "    td['ID'] = td.apply(lambda r: r['CHROM'] + '_' + str(r['POS']) + '_' + str(r['REF']) + '->' + str(r['ALT']), axis=1)\n",
    "    td = td[((td['G70_ref_counts']+td['G70_alt_counts'])>4) & ((td['G70_alt_counts']/(td['G70_ref_counts']+td['G70_alt_counts']))>0.9)]\n",
    "    strain = well_to_strain[well]\n",
    "    for anid in td['ID']:\n",
    "        ids_to_strains_fixed[strain][anid] += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "ids_df = pd.DataFrame({'ID': list(set(list(ids_to_strains_fixed['a'].keys()) + list(ids_to_strains_fixed['alpha'].keys()) + list(ids_to_strains_fixed['diploid'].keys())))})\n",
    "for strain in strains:\n",
    "    numwells = len([w for w in wells if well_to_strain[w]==strain])\n",
    "    ids_df[strain+'_percent_fixed'] = ids_df['ID'].apply(lambda i: ids_to_strains_fixed[strain][i]/numwells)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "ids_to_strains_freqs = {s: defaultdict(list) for s in strains}\n",
    "for well in wells:\n",
    "    td = pd.read_csv('../Output/WGS/combined_option/well_output/'+well + '_full.tsv', delimiter='\\t')\n",
    "    td['ID'] = td.apply(lambda r: r['CHROM'] + '_' + str(r['POS']) + '_' + str(r['REF']) + '->' + str(r['ALT']), axis=1)\n",
    "    td = td[(td['G70_ref_counts']+td['G70_alt_counts'])>4]\n",
    "    td['freq_70'] = td['G70_alt_counts']/(td['G70_ref_counts']+td['G70_alt_counts'])\n",
    "    td = td[td['ID'].isin(list(ids_df['ID']))]\n",
    "    strain = well_to_strain[well]\n",
    "    for anid, freq in np.array(td[['ID', 'freq_70']]):\n",
    "        ids_to_strains_freqs[strain][anid].append(freq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "ids_to_info = dict()\n",
    "for well in wells:\n",
    "    td = pd.read_csv('../Output/WGS/combined_option/well_output/'+well + '_full.tsv', delimiter='\\t')\n",
    "    td['ID'] = td.apply(lambda r: r['CHROM'] + '_' + str(r['POS']) + '_' + str(r['REF']) + '->' + str(r['ALT']), axis=1)\n",
    "    td = td[td['ID'].isin(list(ids_df['ID']))]\n",
    "    for anid, ann, svtype in np.array(td[['ID', 'ANN', 'SVTYPE']]):\n",
    "        ids_to_info[anid] = [ann, svtype]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "for strain in strains:\n",
    "    numwells = len([w for w in wells if well_to_strain[w]==strain])\n",
    "    ids_df[strain+'_mean_freq'] = ids_df['ID'].apply(lambda i: np.sum(ids_to_strains_freqs[strain][i])/numwells)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "ids_df['ANN'] = ids_df['ID'].apply(lambda i: ids_to_info[i][0])\n",
    "ids_df['SVTYPE'] = ids_df['ID'].apply(lambda i: ids_to_info[i][1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "51"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ids_df['summer'] = np.sum(ids_df[[s+'_percent_fixed' for s in strains]], axis=1)\n",
    "ids_df['maxxer'] = np.max(ids_df[[s+'_percent_fixed' for s in strains]], axis=1)\n",
    "ids_df['minner'] = np.min(ids_df[[s+'_percent_fixed' for s in strains]], axis=1)\n",
    "len(ids_df[(ids_df['maxxer']>0.7) & (ids_df['minner']==0)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "td = ids_df[(ids_df['maxxer']>0.7) & (ids_df['minner']==0)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def genes_affected(a):\n",
    "    keep_anns = []\n",
    "    for ann in str(a).split(','):\n",
    "        if '|' in ann:\n",
    "            sa = ann.split('|')\n",
    "            for mut_type in sa[1].split('&'):\n",
    "                if sa[1] not in ['SV', 'synonymous', 'noncoding']:\n",
    "                    if 'Dubious' not in sa[5]:\n",
    "                        keep_anns.append(sa[4])\n",
    "    return ';'.join(keep_anns)\n",
    "\n",
    "def simple_ann(a):\n",
    "    sa = a.split('|')\n",
    "    # format: ALT|simple mutation type|Putative_impact|Gene/ORF name (to be used)|ORF name|proteinAA position|briefDescription\n",
    "    if len(sa) > 13:\n",
    "        aa_pos = sa[13]\n",
    "    else:\n",
    "        aa_pos = ''\n",
    "    return '|'.join([str(i).replace(',', '_') for i in [sa[0], sa[1], sa[2], o2g.get(sa[3], [sa[3]])[0], sa[3], aa_pos, o2g.get(sa[3], [None, sa[3]])[1]]])\n",
    "\n",
    "def annotation_parser(row):\n",
    "    # input is annotation column\n",
    "    # goal is to remove annotations that are \"boring\" - intergenic or more than 100 bp upstream\n",
    "    # simplifies annotations to the format: ALT|simple mutation type|Putative_impact|Gene/ORF name (to be used)|ORF name|proteinAA position|briefDescription\n",
    "    # only preserves the most \"important\" annotation\n",
    "    if pd.notnull(row['SVTYPE']):\n",
    "        return '|SV|||||'\n",
    "    annotations = row['ANN']\n",
    "    anns = []\n",
    "    for a in str(annotations).split(','):\n",
    "        sa = a.split('|')\n",
    "        if len(sa) > 1:\n",
    "            effs = sa[1]\n",
    "            # excluding intergenic or downstream annotations\n",
    "            if effs not in ['intergenic_region', 'downstream_gene_variant']:\n",
    "                if 'upstream' in effs and len(sa) > 14: # if upstream, must be within 100 bp of the feature to keep it\n",
    "                    distance_to_feature = sa[14]\n",
    "                    if distance_to_feature != '':\n",
    "                        if int(distance_to_feature) < 100:\n",
    "                            anns.append(simple_ann(a))\n",
    "                else:\n",
    "                    anns.append(simple_ann(a))\n",
    "    return ','.join(anns)\n",
    "\n",
    "\n",
    "\n",
    "td['ANN_simpler'] = td.apply(lambda r: annotation_parser(r), axis=1)\n",
    "td['ORF_hit'] = td['ANN_simpler'].apply(lambda a: genes_affected(a))\n",
    "td['Gene_hit'] = td['ORF_hit'].apply(lambda o: o2g.setdefault(o, o))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "td.to_csv('strain_difs_rough.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Gene_hit</th>\n",
       "      <th>diploid_mean_freq</th>\n",
       "      <th>alpha_mean_freq</th>\n",
       "      <th>a_mean_freq</th>\n",
       "      <th>ANN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>chrX_519278_G-&gt;A</td>\n",
       "      <td></td>\n",
       "      <td>0.510897</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.999369</td>\n",
       "      <td>A|upstream_gene_variant|MODIFIER|YJR043C|YJR04...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>374</th>\n",
       "      <td>chrX_354268_G-&gt;A</td>\n",
       "      <td></td>\n",
       "      <td>0.516654</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.996528</td>\n",
       "      <td>A|upstream_gene_variant|MODIFIER|YJL047C|YJL04...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>417</th>\n",
       "      <td>chrXI_492323_A-&gt;G</td>\n",
       "      <td>BCH2</td>\n",
       "      <td>0.489544</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.999042</td>\n",
       "      <td>G|missense_variant|MODERATE|YKR027W|YKR027W|tr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>443</th>\n",
       "      <td>chrXVI_449037_C-&gt;A</td>\n",
       "      <td>PDR12</td>\n",
       "      <td>0.497349</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.998148</td>\n",
       "      <td>A|stop_gained|HIGH|YPL058C|YPL058C|transcript|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>517</th>\n",
       "      <td>chrXIII_220613_G-&gt;A</td>\n",
       "      <td>TSA1</td>\n",
       "      <td>0.490381</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>A|missense_variant|MODERATE|YML028W|YML028W|tr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>557</th>\n",
       "      <td>chrXIV_68875_C-&gt;A</td>\n",
       "      <td></td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.996876</td>\n",
       "      <td>A|upstream_gene_variant|MODIFIER|YNL301C|YNL30...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>567</th>\n",
       "      <td>chrXII_878086_G-&gt;A</td>\n",
       "      <td></td>\n",
       "      <td>0.478287</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.998889</td>\n",
       "      <td>A|upstream_gene_variant|MODIFIER|YLR376C|YLR37...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>580</th>\n",
       "      <td>chrXV_954123_G-&gt;C</td>\n",
       "      <td>TEA1</td>\n",
       "      <td>0.487222</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.996693</td>\n",
       "      <td>C|missense_variant|MODERATE|YOR337W|YOR337W|tr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>653</th>\n",
       "      <td>chrXII_940291_T-&gt;A</td>\n",
       "      <td>VIP1</td>\n",
       "      <td>0.476443</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.996554</td>\n",
       "      <td>A|missense_variant|MODERATE|YLR410W|YLR410W|tr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>701</th>\n",
       "      <td>chrII_109576_A-&gt;C</td>\n",
       "      <td>YEL1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.994936</td>\n",
       "      <td>C|missense_variant|MODERATE|YBL060W|YBL060W|tr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>730</th>\n",
       "      <td>chrXV_428975_A-&gt;T</td>\n",
       "      <td>VHS3</td>\n",
       "      <td>0.482295</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.995769</td>\n",
       "      <td>T|missense_variant|MODERATE|YOR054C|YOR054C|tr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>732</th>\n",
       "      <td>chrXII_12088_T-&gt;G</td>\n",
       "      <td></td>\n",
       "      <td>0.506667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.996558</td>\n",
       "      <td>G|upstream_gene_variant|MODIFIER|YLL066C|YLL06...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>740</th>\n",
       "      <td>chrVII_115328_T-&gt;G</td>\n",
       "      <td></td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.996841</td>\n",
       "      <td>G|upstream_gene_variant|MODIFIER|YGL204C|YGL20...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>763</th>\n",
       "      <td>chrVII_27252_T-&gt;G</td>\n",
       "      <td>RTG2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.994910</td>\n",
       "      <td>G|missense_variant|MODERATE|YGL252C|YGL252C|tr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>782</th>\n",
       "      <td>chrIV_719939_AACT-&gt;A</td>\n",
       "      <td>MRX16</td>\n",
       "      <td>0.472994</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.998016</td>\n",
       "      <td>A|conservative_inframe_deletion|MODERATE|YDR13...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>822</th>\n",
       "      <td>chrXIV_193606_C-&gt;T</td>\n",
       "      <td>ATG2</td>\n",
       "      <td>0.504030</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.994849</td>\n",
       "      <td>T|synonymous_variant|LOW|YNL242W|YNL242W|trans...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>987</th>\n",
       "      <td>chrIV_786229_C-&gt;A</td>\n",
       "      <td>SEC5</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.993660</td>\n",
       "      <td>A|missense_variant|MODERATE|YDR166C|YDR166C|tr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>992</th>\n",
       "      <td>chrXIII_371867_TTCG-&gt;T</td>\n",
       "      <td></td>\n",
       "      <td>0.479723</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.996643</td>\n",
       "      <td>T|upstream_gene_variant|MODIFIER|YMR049C|YMR04...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1063</th>\n",
       "      <td>chrIV_1346602_C-&gt;T</td>\n",
       "      <td>SSN2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.991352</td>\n",
       "      <td>T|missense_variant|MODERATE|YDR443C|YDR443C|tr...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          ID Gene_hit  diploid_mean_freq  alpha_mean_freq  \\\n",
       "259         chrX_519278_G->A                    0.510897              0.0   \n",
       "374         chrX_354268_G->A                    0.516654              0.0   \n",
       "417        chrXI_492323_A->G     BCH2           0.489544              0.0   \n",
       "443       chrXVI_449037_C->A    PDR12           0.497349              0.0   \n",
       "517      chrXIII_220613_G->A     TSA1           0.490381              0.0   \n",
       "557        chrXIV_68875_C->A                    0.000000              0.0   \n",
       "567       chrXII_878086_G->A                    0.478287              0.0   \n",
       "580        chrXV_954123_G->C     TEA1           0.487222              0.0   \n",
       "653       chrXII_940291_T->A     VIP1           0.476443              0.0   \n",
       "701        chrII_109576_A->C     YEL1           0.000000              0.0   \n",
       "730        chrXV_428975_A->T     VHS3           0.482295              0.0   \n",
       "732        chrXII_12088_T->G                    0.506667              0.0   \n",
       "740       chrVII_115328_T->G                    0.000000              0.0   \n",
       "763        chrVII_27252_T->G     RTG2           0.000000              0.0   \n",
       "782     chrIV_719939_AACT->A    MRX16           0.472994              0.0   \n",
       "822       chrXIV_193606_C->T     ATG2           0.504030              0.0   \n",
       "987        chrIV_786229_C->A     SEC5           0.000000              0.0   \n",
       "992   chrXIII_371867_TTCG->T                    0.479723              0.0   \n",
       "1063      chrIV_1346602_C->T     SSN2           0.000000              0.0   \n",
       "\n",
       "      a_mean_freq                                                ANN  \n",
       "259      0.999369  A|upstream_gene_variant|MODIFIER|YJR043C|YJR04...  \n",
       "374      0.996528  A|upstream_gene_variant|MODIFIER|YJL047C|YJL04...  \n",
       "417      0.999042  G|missense_variant|MODERATE|YKR027W|YKR027W|tr...  \n",
       "443      0.998148  A|stop_gained|HIGH|YPL058C|YPL058C|transcript|...  \n",
       "517      1.000000  A|missense_variant|MODERATE|YML028W|YML028W|tr...  \n",
       "557      0.996876  A|upstream_gene_variant|MODIFIER|YNL301C|YNL30...  \n",
       "567      0.998889  A|upstream_gene_variant|MODIFIER|YLR376C|YLR37...  \n",
       "580      0.996693  C|missense_variant|MODERATE|YOR337W|YOR337W|tr...  \n",
       "653      0.996554  A|missense_variant|MODERATE|YLR410W|YLR410W|tr...  \n",
       "701      0.994936  C|missense_variant|MODERATE|YBL060W|YBL060W|tr...  \n",
       "730      0.995769  T|missense_variant|MODERATE|YOR054C|YOR054C|tr...  \n",
       "732      0.996558  G|upstream_gene_variant|MODIFIER|YLL066C|YLL06...  \n",
       "740      0.996841  G|upstream_gene_variant|MODIFIER|YGL204C|YGL20...  \n",
       "763      0.994910  G|missense_variant|MODERATE|YGL252C|YGL252C|tr...  \n",
       "782      0.998016  A|conservative_inframe_deletion|MODERATE|YDR13...  \n",
       "822      0.994849  T|synonymous_variant|LOW|YNL242W|YNL242W|trans...  \n",
       "987      0.993660  A|missense_variant|MODERATE|YDR166C|YDR166C|tr...  \n",
       "992      0.996643  T|upstream_gene_variant|MODIFIER|YMR049C|YMR04...  \n",
       "1063     0.991352  T|missense_variant|MODERATE|YDR443C|YDR443C|tr...  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "td[(td['a_percent_fixed']==1) & (td['summer']==1)][['ID', 'Gene_hit'] + [s + '_mean_freq' for s in strains] + ['ANN']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TSA1 could cause a mutation rate increase: https://www.yeastgenome.org/reference/S000130731"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "milo_py37",
   "language": "python",
   "name": "milo_py37"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
